soci-snapshotter: use socket activation, embed build commit

ginglis13 opened the pull request:
<!--
Tips:
- Please read CONTRIBUTING.md to understand our process and our requests for PRs.
- Please file an issue before creating a PR so we can discuss the change and confirm it's not already being worked on.
-->

**Issue number:**

Closes #530 

**Description of changes:**

Start the soci-snapshotter service with systemd socket activation and use its git revision as build metadata embedded in binaries


**Testing done:**

Testing `gitrev` on an `aws-dev` variant:

```
bash-5.1# soci-snapshotter-grpc --version
soci-snapshotter-grpc version v0.9.0+bottlerocket 737f61a3db40c386f997c1f126344158aa3ad43c
```

Testing socket activation on `aws-dev` variant with containerd configured to load soci:

```
bash-5.1# systemctl status soci-snapshotter
● soci-snapshotter.service - soci snapshotter containerd plugin
     Loaded: loaded (/x86_64-bottlerocket-linux-gnu/sys-root/usr/lib/systemd/system/soci-snapshotter.service; enabled; preset: enabled)
    Drop-In: /x86_64-bottlerocket-linux-gnu/sys-root/usr/lib/systemd/system/service.d
             └─00-aws-config.conf
     Active: active (running) since Fri 2025-05-30 00:53:42 UTC; 8min ago
TriggeredBy: ● soci-snapshotter.socket
       Docs: https://github.com/awslabs/soci-snapshotter
   Main PID: 1230 (soci-snapshotte)
      Tasks: 40 (limit: 8934)
     Memory: 717.8M
        CPU: 6.301s
     CGroup: /system.slice/soci-snapshotter.service
             └─1230 /usr/bin/soci-snapshotter-grpc --address fd://
```

Configured containerd:

```
bash-5.1# tail -n 5 /etc/containerd/config.toml

[proxy_plugins]
  [proxy_plugins.soci]
    type = "snapshot"
    address = "/run/soci-snapshotter-grpc/soci-snapshotter-grpc.sock"
```

Pulled example image:

```
ctr i pull --snapshotter=soci public.ecr.aws/soci-workshop-examples/ffmpeg:latest
```

On an `aws-dev` instance without containerd configured with soci:

```
bash-5.1# systemctl status soci-snapshotter
○ soci-snapshotter.service - soci snapshotter containerd plugin
     Loaded: loaded (/x86_64-bottlerocket-linux-gnu/sys-root/usr/lib/systemd/system/soci-snapshotter.service; static)
    Drop-In: /x86_64-bottlerocket-linux-gnu/sys-root/usr/lib/systemd/system/service.d
             └─00-aws-config.conf
     Active: inactive (dead)
TriggeredBy: ● soci-snapshotter.socket
       Docs: https://github.com/awslabs/soci-snapshotter
```

**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

ginglis13 committed:
soci-snapshotter: use socket activation

Start the soci-snapshotter service with systemd socket activation such
that it only activates and runs when traffic is sent to the snapshotter.
This allows for more dynamic configuration of containerd without using
unnecessary system resources.

Signed-off-by: Gavin Inglis <giinglis@amazon.com>

bcressey commented:
In lieu of a related issue, can you split this into two commits, and add a 1-2 sentence description to each commit on why the change is useful or desirable?

Right now everything describes "what" is changing but not "why".

ginglis13 commented:
> In lieu of a related issue, can you split this into two commits, and add a 1-2 sentence description to each commit on why the change is useful or desirable?

I've just created #530 but +1 to splitting out commits, can do

ginglis13 committed:
soci-snapshotter: embed git revision and version

Follow the practice of other Go binaries packaged for Bottlerocket by
including the versioning and git commit information of the package in
provided binaries.

Signed-off-by: Gavin Inglis <giinglis@amazon.com>
